### PR TITLE
Ledger TX Modals

### PR DESCRIPTION
**Pull request by GummyGod**
**Description**

_Write a short description of the changes included in this pull request._

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

**Issue**

Progresses #???

### CODE REVIEW

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -8,7 +8,6 @@
 
 /* Begin PBXBuildFile section */
 		07CBADD9D4B441008304F8D3 /* EuclidCircularB-Light.otf in Resources */ = {isa = PBXBuildFile; fileRef = A98029A3662F4C1391489A6B /* EuclidCircularB-Light.otf */; };
-		0FD509E0336BF221F6527B24 /* libPods-MetaMask.a in Frameworks */ = {isa = PBXBuildFile; fileRef = EA0FF04C201CE95ED7EEEA82 /* libPods-MetaMask.a */; };
```
Comment (zoltanarvai) [inline ios/MetaMask.xcodeproj/project.pbxproj:11]: you need to revert the changes on this file

File: ios/MetaMask/MetaMaskDebug.entitlements [hunk 2]
```diff
@@ -1,19 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
-<dict>
```
Comment (zoltanarvai) [inline ios/MetaMask/MetaMaskDebug.entitlements:4]: revert the changes here

File: ios/Podfile [hunk 3]
```diff
@@ -37,7 +37,8 @@ target 'MetaMask' do
   config = use_native_modules!
   # Comment the next line if you don't want to use dynamic frameworks
   # use_frameworks!
-	pod 'Permission-BluetoothPeripheral', :path => '../node_modules/react-native-permissions/ios/BluetoothPeripheral'
```
Comment (zoltanarvai) [inline ios/Podfile:40]: this change is not necessary

File: ios/MetaMask/Info.plist [hunk 4]
```diff
@@ -128,5 +128,13 @@
 	<string>$(MM_FOX_CODE)</string>
 	<key>mixpanel_token</key>
 	<string>$(MM_MIXPANEL_TOKEN)</string>
+	<key>NSBluetoothPeripheralUsageDescription</key>
+  <string>Metamask needs Bluetooth access to connect to external devices.</string>
```
Comment (zoltanarvai) [inline ios/MetaMask/Info.plist:132]: you should not have these here

File: app/reducers/modals/index.js [hunk 5]
```diff
@@ -54,6 +56,17 @@ const modalsReducer = (state = initialState, action) => {
 				...state,
 				approveModalVisible: !state.approveModalVisible,
 			};
+		case 'TOGGLE_LEDGER_TRANSACTION':
```
Comment (zoltanarvai) [inline app/reducers/modals/index.js:59]: shall we call this device action? (can we use this same modal for transactions, signing messages, signing typed data?)

File: app/components/Views/LedgerConnect/index.tsx [hunk 6]
```diff
@@ -82,10 +94,12 @@ const LedgerConnect = () => {
 		// Establish bluetooth connection to ledger
 		try {
 			if (!transportRef.current && selectedDevice) {
+				console.log('selDev', selectedDevice);
 				transportRef.current = await BluetoothTransport.open(selectedDevice);
 				transportRef.current?.on('disconnect', () => (transportRef.current = undefined));
 			}
-		} catch {
+		} catch (e) {
+			console.log('err,', e);
```
Comment (zoltanarvai) [inline app/components/Views/LedgerConnect/index.tsx:102]: remove console.logs pls

File: app/components/Views/LedgerConnect/index.tsx [hunk 7]
```diff
@@ -141,6 +155,20 @@ const LedgerConnect = () => {
 				<View style={styles.textContainer}>
 					<Text bold>{strings('ledger.looking_for_device')}</Text>
 					<Text style={styles.instructionsText}>{strings('ledger.ledger_reminder_message')}</Text>
+					<TouchableOpacity
```
Comment (zoltanarvai) [inline app/components/Views/LedgerConnect/index.tsx:158]: why is this thing here?

File: app/components/Views/LedgerConnect/Scan.tsx [hunk 8]
```diff
@@ -66,7 +66,7 @@ const Scan = ({ onDeviceSelected }: { onDeviceSelected: (device: NanoDevice) =>
 	useEffect(() => {
 		const run = async () => {
 			if (Device.isIos()) {
-				const bluetoothPermissionStatus = await check(PERMISSIONS.IOS.BLUETOOTH_PERIPHERAL);
+				const bluetoothPermissionStatus = await request(PERMISSIONS.IOS.BLUETOOTH_PERIPHERAL);
```
Comment (zoltanarvai) [inline app/components/Views/LedgerConnect/Scan.tsx:69]: probably no need for this change

File: app/components/UI/Tokens/index.js [hunk 9]
```diff
@@ -251,10 +251,10 @@ class Tokens extends PureComponent {
 		const { tokens, hideZeroBalanceTokens, tokenBalances } = this.props;
 		const tokensToDisplay = hideZeroBalanceTokens
 			? tokens.filter((token) => {
-					const { address, isETH } = token;
-					return !isZero(tokenBalances[address]) || isETH;
```
Comment (zoltanarvai) [inline app/components/UI/Tokens/index.js:255]: why change this?

File: app/components/UI/LedgerModals/LedgerTransactionModal.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		activityIndicatorContainer: {
+			marginTop: 35,
+		},
+		buttonContainer: {
+			marginTop: 35,
+			width: '90%',
+		},
+	});
+
+const ledgerConnectImage = require('../../../images/ledger-connect.png');
```
Comment (zoltanarvai) [inline app/components/UI/LedgerModals/LedgerTransactionModal.tsx:35]: we need two images... one for DARK MODE and one for LIGHT MODE

File: app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+			width: '90%',
+			flexDirection: 'row',
+			justifyContent: 'space-evenly',
+		},
+		buttonStyle: {
+			width: Device.getDeviceWidth() * 0.3,
+		},
+	});
+
+const ledgerConnectImage = require('../../../images/ledger-connect.png');
```
Comment (zoltanarvai) [inline app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx:41]: see my comment blow for images

File: app/components/UI/LedgerModals/InfoBox.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import { StyleSheet, View } from 'react-native';
+import Text from '../../Base/Text';
+import { mockTheme, useAppThemeFromContext } from '../../../util/theme';
+import { Colors } from '../../../util/theme/models';
+import { strings } from '../../../../locales/i18n';
+
+const createStyles = (colors: Colors) =>
+	StyleSheet.create({
+		infoBoxContainer: {
+			backgroundColor: '#367bcf14',
```
Comment (zoltanarvai) [inline app/components/UI/LedgerModals/InfoBox.tsx:11]: no already existing colour for this one?

**Review by zoltanarvai — CHANGES_REQUESTED** [review]

File: app/components/UI/LedgerModals/InfoBox.tsx (hunk 12 above)
Comment (GummyGod) [reply app/components/UI/LedgerModals/InfoBox.tsx:11]: No color, this one is pretty custom

File: app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx (hunk 11 above)
Comment (GummyGod) [reply app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx:41]: Same here, I kept the same image, i think it should be fine til we get the two versions

File: app/components/UI/Tokens/index.js (hunk 9 above)
Comment (GummyGod) [reply app/components/UI/Tokens/index.js:255]: linter

File: app/components/Views/LedgerConnect/Scan.tsx (hunk 8 above)
Comment (GummyGod) [reply app/components/Views/LedgerConnect/Scan.tsx:69]: you suggested we keep request

File: app/components/Views/LedgerConnect/index.tsx (hunk 7 above)
Comment (GummyGod) [reply app/components/Views/LedgerConnect/index.tsx:158]: It is a way for me to test the modal's UI until we add the real logic. I'll comment it

File: ios/MetaMask/Info.plist (hunk 4 above)
Comment (GummyGod) [reply ios/MetaMask/Info.plist:132]: why?

File: ios/Podfile (hunk 3 above)
Comment (GummyGod) [reply ios/Podfile:40]: fixed

File: ios/MetaMask/Info.plist (hunk 4 above)
Comment (zoltanarvai) [reply ios/MetaMask/Info.plist]: they should be coming from the main project... so not sure why you have them.

File: app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+			<View style={styles.contentWrapper}>
+				<Image source={ledgerConnectImage} />
+				<View style={styles.confirmTransactionTextContainer}>
+					<Text bold big>
+						{strings('ledger.confirm_transaction_on_ledger')}
+					</Text>
+				</View>
+				<InfoBox />
+				<View style={styles.buttonContainer}>
+					<StyledButton type="cancel" onPress={() => null} style={styles.buttonStyle}>
```
Comment (fstream9) [inline app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx:58]: shouldn't these buttons do something?

File: app/components/UI/LedgerModals/LedgerTransactionModal.tsx [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+					<Text bold big>
+						{strings('ledger.confirm_transaction_on_ledger')}
+					</Text>
+				</View>
+				<View style={styles.activityIndicatorContainer}>
+					<ActivityIndicator />
+				</View>
+				<InfoBox />
+				<View style={styles.buttonContainer}>
+					<StyledButton type="cancel" onPress={() => null}>
```
Comment (fstream9) [inline app/components/UI/LedgerModals/LedgerTransactionModal.tsx:55]: same here, do something?

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 15]
```diff
@@ -821,7 +821,7 @@
 				CODE_SIGN_ENTITLEMENTS = MetaMask/MetaMaskDebug.entitlements;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				CODE_SIGN_STYLE = Manual;
-				CURRENT_PROJECT_VERSION = 859;
+				CURRENT_PROJECT_VERSION = 875;
```
Comment (fstream9) [inline ios/MetaMask.xcodeproj/project.pbxproj:824]: quite a big jump?

File: app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx (hunk 13 above)
Comment (GummyGod) [reply app/components/UI/LedgerModals/LedgerTransactionFailModal.tsx:58]: Yes, but not atm, this is the MVP for them, we need to refactor first

File: app/components/UI/LedgerModals/LedgerTransactionModal.tsx (hunk 14 above)
Comment (GummyGod) [reply app/components/UI/LedgerModals/LedgerTransactionModal.tsx:55]: same

File: ios/MetaMask.xcodeproj/project.pbxproj (hunk 15 above)
Comment (GummyGod) [reply ios/MetaMask.xcodeproj/project.pbxproj:824]: That was automatic

Comment (zoltanarvai) [reply ios/MetaMask.xcodeproj/project.pbxproj:824]: I never commit these in.... not sure what this does. Better not update?

File: app/actions/modals/index.js [hunk 16]
```diff
@@ -37,3 +37,11 @@ export function toggleApproveModal(show) {
 		show,
 	};
 }
+
+export const toggleLedgerTransactionModal = () => ({
+	type: 'TOGGLE_LEDGER_DEVICE_TRANSACTION',
```
Comment (zoltanarvai) [inline app/actions/modals/index.js:42]: We were saying that this is not about transaction. It's a generic DeviceAction.

We might use it for signing, transaction, personal message signing, typed data, etc.

I'd prefer to use `TOGGLE_LEDGER_DEVICE_ACTION`

and toggleLedgerDeviceActionModal

File: app/components/UI/LedgerModals/LedgerTransactionModal.tsx (hunk 10 above)
Comment (zoltanarvai) [reply app/components/UI/LedgerModals/LedgerTransactionModal.tsx:35]: Can you ask Antonella to provide this? She has actually two of these images. (one with a bluetooth sign in the middle and one with a question mark)

File: app/components/Views/LedgerConnect/index.tsx [hunk 17]
```diff
@@ -141,6 +153,20 @@ const LedgerConnect = () => {
 				<View style={styles.textContainer}>
 					<Text bold>{strings('ledger.looking_for_device')}</Text>
 					<Text style={styles.instructionsText}>{strings('ledger.ledger_reminder_message')}</Text>
+					{/* <TouchableOpacity
```
Comment (zoltanarvai) [inline app/components/Views/LedgerConnect/index.tsx:156]: we can safely remove this for now

File: ios/MetaMask.xcodeproj/project.pbxproj [hunk 18]
```diff
@@ -923,7 +923,7 @@
 					"\"$(SRCROOT)/MetaMask/System/Library/Frameworks\"",
 				);
 				LLVM_LTO = YES;
-				MARKETING_VERSION = 4.3.1;
+				MARKETING_VERSION = 5.0.0;
```
Comment (zoltanarvai) [inline ios/MetaMask.xcodeproj/project.pbxproj:926]: definitely don't update this version please yourself

File: app/components/Views/LedgerConnect/index.tsx [hunk 19]
```diff
@@ -5,15 +5,14 @@ import { useNavigation } from '@react-navigation/native';
...
-
 import Engine from '../../../core/Engine';
 import StyledButton from '../../../components/UI/StyledButton';
 import Text from '../../../components/Base/Text';
 import { mockTheme, useAppThemeFromContext } from '../../../util/theme';
 import Device from '../../../util/device';
 import { fontStyles } from '../../../styles/common';
 import Scan from './Scan';
-
+import { useDispatch } from 'react-redux';
```
Comment (zoltanarvai) [inline app/components/Views/LedgerConnect/index.tsx:15]: this should be on the top as an import. (before local imports)

File: app/reducers/modals/index.js [hunk 20]
```diff
@@ -6,6 +6,8 @@ const initialState = {
 	receiveAsset: undefined,
 	dappTransactionModalVisible: false,
 	approveModalVisible: false,
+	ledgerTransactionModalVisible: false,
```
Comment (zoltanarvai) [inline app/reducers/modals/index.js:9]: you didn't change the naming here

File: ios/MetaMask.xcodeproj/project.pbxproj (hunk 15 above)
Comment (zoltanarvai) [reply ios/MetaMask.xcodeproj/project.pbxproj:824]: these changes are still here

**Review by zoltanarvai — APPROVED** [review]

**Pull request merged**